File: 3. CounterApp/app/src/main/java/com/mo/counterapp/MainActivity.java
package com.mo.counterapp;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {


    private TextView textView;
    private Button button;

    private int count = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        button = findViewById(R.id.button);
        textView = findViewById(R.id.textView);

        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // count = count + 1;
                textView.setText(String.valueOf(changeCounter()));
            }
        });





    }

    public int changeCounter(){
        //count = count + 1;
        return ++count;


    }
}

File: __Kotlin Apps/Corontines/app/src/main/java/com/example/corontinesapp/MainActivity.kt
package com.example.corontinesapp

import android.annotation.SuppressLint
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.withCreated
import com.example.corontinesapp.databinding.ActivityMainBinding
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch


class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding

    private var counter:Int = 0
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = DataBindingUtil.setContentView(this,R.layout.activity_main)

        binding.apply {

            button.setOnClickListener(View.OnClickListener {
                text1.text = counter++.toString()


            })


            button2.setOnClickListener(View.OnClickListener {

                //with using coroutines
                //Dispatchers help coroutines in deciding the thread on which the work has to be done.

                CoroutineScope(Dispatchers.IO).launch {


                    //2
                    //runDownloadBigFileFromServer()

                    //3
                    //Switching between coroutines
                    runDownloadBigAndShowingDataIntoMainThread()

                }


                //1
                /* //without using coroutines
                runDownloadBigFileFromServer()*/





            })

        }


    }


    private fun runDownloadBigFileFromServer() {

        for (i in 1..100000){
            Log.i("TAGY","Downloading $i in ${Thread.currentThread().name}")
        }

    }



    
    private suspend fun runDownloadBigAndShowingDataIntoMainThread() {

        for (i in 1..100000){
            Log.i("TAGY","Downloading $i in ${Thread.currentThread().name}")

            withCreated {
                binding.text2.text = "$i in ${Thread.currentThread().name}"

            }
        }

    }
}